### PR TITLE
Replace TravisCI with Github Actions

### PR DESCRIPTION
**Pull request by oniofchaos**
Closes https://github.com/devise-security/devise-security/issues/102

When cached, the entire matrix of 30 test combinations runs in just under 3 minutes.

Three things I want to iterate on after this:

1) Getting inline Rubocop comments through Github Actions instead of Code Climate, for example by using https://github.com/marketplace/actions/rubocop-checks or https://github.com/andrewmcodes/rubocop-linter-action

2) ~Investigate migrating off Coveralls. It is really built for using with TravisCI and it's not so great outside of that. Need to look for a replacement that's good for Github Actions, or figure out how to use https://github.com/marketplace/actions/coveralls-github-action and https://github.com/simplecov-ruby/simplecov#merging-results properly. It functions right now (I think) though.~

~Was able to figure out how to do this. Just need to figure out how to get Coveralls to leave a comment with the summary.~

~Now Coveralls is cranky about the %~

OK this is working just fine now. Each build has an associated link within Coveralls, and Coveralls also is able to calculate an aggregate percentage.

3) Buff up the aggregated test suite coverage to 100%, so that #2 is easier.

4) Make repo work with `ruby-head`

You can validate that the caching works by seeing the last two builds on my fork:

[uncached](https://github.com/oniofchaos/devise-security/actions/runs/321140628)
[cached](https://github.com/oniofchaos/devise-security/actions/runs/321151631)

### CODE REVIEW

**Comment by tmr08c** [comment]
> Getting inline Rubocop comments through Github Actions instead of Code Climate

I like this idea, but have been less sure about it since the author of one of the Actions [recommends against using it](https://github.com/andrewmcodes/rubocop-linter-action#introduction):

> I would truly recommend not using this action. The potential for random breaking changes from GitHub will always be a looming threat and I have ascertained people are using this action as a crutch to avoid what they really ought be doing.
>
> If you find yourself constantly having to wade through RuboCop errors in GitHub, the fix is to not allow them to get to GitHub broken in the first place. I would highly suggest using a tool like lefthook or similar to flag or fix failing cops before they get to GitHub Actions, saving you and your team time and removing the need for this action.

Git hooks can be nice, but I think they often annoy people more. I sort of like the idea of the Action calling out concerns in the PR. Maybe this means we go with the first Action you linked since that author hasn't said they don't plan to keep up with the Action (though they haven't upated it in 2 years) 🤷 .

**Comment by oniofchaos** [comment]
Yup I agree! I think looking into a custom action would be the best approach as well. I didn't want to start doing that until there was consensus that the Actions approach in general for CI was something that worked.

Once this is approved in merged, that's one of the things I want to look into.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        parallel: true
+        path-to-lcov: ./coverage/lcov/devise-security.lcov
+      if: ${{ matrix.rails != '4.2' || (matrix.rails == '4.2' && matrix.ruby == '2.5') }}
+
+  finish:
+    needs: run-tests
+    runs-on: ubuntu-latest
+    steps:
+    - name: Coveralls Finished
+      uses: coverallsapp/github-action@master
```
Comment (tmr08c) [inline .github/workflows/main.yml:60]: Based on their [docs](https://github.com/marketplace/actions/coveralls-github-action), it looks like this Action should add a comment with test coverage information so it seems like we shouldn't miss out on anything from Coveralls 👍  (not sure if we did anything with checking for a certain percentage of coverage).

**Review by tmr08c — APPROVED** [review]
I am interested in the idea of moving towards Actions, so I am excited about this. However, I will admit I am excited by the "new, shiny" and I am not being particularly pragmatic about why we _should_ stay with Travis. 

It seems way faster which seems like a big win.

I also wonder if it's easier for contributors to work with - can a contributor for an Action to re-run if there is a flaky test? I couldn't find a way to tell Travis to re-run, which was a bit of a pain when we had some flaky test issues. 

Some additional thoughts/questions:

* If we do move forward with this we can look into replacing the README's Travis badge with a [GitHub Actions one](https://docs.github.com/en/free-pro-team@latest/actions/managing-workflow-runs/adding-a-workflow-status-badge).
* Do we want to find the equivalent of something like `wwtd` for local testing? A quick Google search found [act](https://github.com/nektos/act) which looks like it would run the whole workflow locally and not a particular matrix combination. We probably don't have to have an answer now, but may want to update the README to remove `wwtd` if we start ripping out Travis config.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    branches: [ master ]
+jobs:
+  run-tests:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        orm: [active_record, mongoid]
+        rails: ['4.2', '5.0', '5.1', '5.2', '6.0']
+        ruby: ['2.5', '2.6', '2.7'] # TODO fix repo to work with ruby-head
```
[the remark this answers is not in the document]
Comment (oniofchaos) [reply .github/workflows/main.yml:16]: Mind if I do this in a follow-up PR?

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      with:
+        path: |
+          **/bundle
+        key: bundle-use-ruby-${{ matrix.ruby }}-rails-${{ matrix.rails }}-orm-${{ matrix.orm }}-gemfile-${{ hashFiles(format('**/gemfiles/rails_{0}.gemfile', matrix.rails)) }}-gemspec-${{ hashFiles('**/devise-security.gemspec') }}
+    - name: bundle install
+      env:
+        BUNDLE_GEMFILE: gemfiles/rails_${{ matrix.rails }}.gemfile
+      run: |
+        bundle config path vendor/bundle
+        bundle install --jobs 4
```
[the remark this answers is not in the document]
Comment (oniofchaos) [reply .github/workflows/main.yml:33]: I'll just go ahead and remove the jobs arg and let bundler handle it internally.

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      if: ${{ matrix.rails != '4.2' || (matrix.rails == '4.2' && matrix.ruby == '2.5') }}
```
[the remark this answers is not in the document]
Comment (oniofchaos) [reply .github/workflows/main.yml:34]: Agreed, this is likely possible. Also mind if I do it in a separate PR? :)

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        flag-name: run-orm-${{ matrix.orm }}-rails-${{ matrix.rails }}-ruby-${{ matrix.ruby }}
+        parallel: true
+        path-to-lcov: ./coverage/lcov/devise-security.lcov
+      if: ${{ matrix.rails != '4.2' || (matrix.rails == '4.2' && matrix.ruby == '2.5') }}
+
+  finish:
+    needs: run-tests
+    runs-on: ubuntu-latest
+    steps:
+    - name: Coveralls Finished
```
[the remark this answers is not in the document]
Comment (oniofchaos) [reply .github/workflows/main.yml:59]: Agreed.

File: .github/workflows/main.yml (hunk 1 above)
Comment (oniofchaos) [reply .github/workflows/main.yml:60]: https://github.com/devise-security/devise-security/pull/252#issuecomment-714220480 that's what happened here? Unless you're referring to something else?

**Review by oniofchaos — COMMENTED** [review]
> I am interested in the idea of moving towards Actions, so I am excited about this. However, I will admit I am excited by the "new, shiny" and I am not being particularly pragmatic about why we should stay with Travis.

TravisCI was gutted about a year ago. See https://news.ycombinator.com/item?id=19218036. I don't think it's a good idea to use a tool that its owners aren't investing in.

Additionally, migrating to Github Actions keeps a major part of the SLDC for this gem within Github. Fewer systems == better as a general rule IMO. Github is continuing to iterate on it as well.

Finally, I really didn't like the UI of TravisCI. That's just a personal issue though :)

> It seems way faster which seems like a big win.

That was honestly the biggest driving force for me doing this. The wait times after merging in latest master were making me so frustrated.

> I also wonder if it's easier for contributors to work with - can a contributor for an Action to re-run if there is a flaky test? I couldn't find a way to tell Travis to re-run, which was a bit of a pain when we had some flaky test issues.

https://github.com/devise-security/devise-security/actions/runs/321299054 If you go to an individual Action result, there's a re-run jobs button in the top right.

> Some additional thoughts/questions:
> If we do move forward with this we can look into replacing the README's Travis badge with a GitHub Actions one.

Definitely. Will do a follow-up on this.

> Do we want to find the equivalent of something like wwtd for local testing? A quick Google search found act which looks like it would run the whole workflow locally and not a particular matrix combination. We probably don't have to have an answer now, but may want to update the README to remove wwtd if we start ripping out Travis config.

https://github.com/nektos/act looks like this could be an option to replace.  Thoughts?

File: .github/workflows/main.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (oniofchaos) [reply .github/workflows/main.yml:33]:
```suggestion
        bundle install
```

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        parallel: true
+        path-to-lcov: ./coverage/lcov/devise-security.lcov
+      if: ${{ matrix.rails != '4.2' || (matrix.rails == '4.2' && matrix.ruby == '2.5') }}
+
+  finish:
+    needs: run-tests
+    runs-on: ubuntu-latest
+    steps:
+    - name: Upload coverage to Coveralls
+      uses: coverallsapp/github-action@master
```
Comment (oniofchaos) [inline .github/workflows/main.yml:60]:
```suggestion
      uses: coverallsapp/github-action@v1.1.2
```

File: .github/workflows/main.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (tmr08c) [reply .github/workflows/main.yml]: That makes sense 👍

File: .github/workflows/main.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (tmr08c) [reply .github/workflows/main.yml]: That makes sense 👍

**Pull request merged**